consensus/cometbft/service/abci.go

```diff
@@ -143,11 +147,88 @@ func (s *Service) Commit(
 // NOOP methods
 //
 
-func (Service) Query(
-	context.Context,
-	*abci.QueryRequest,
+func (s *Service) Query(
+	_ context.Context,
+	req *abci.QueryRequest,
 ) (*abci.QueryResponse, error) {
-	return &abci.QueryResponse{}, nil
+	resp := &abci.QueryResponse{}
+
+	// add panic recovery for all queries
+	//
+	// Ref: https://github.com/cosmos/cosmos-sdk/pull/8039
+	defer func() {
+		if r := recover(); r != nil {
+			resp = queryResult(errorsmod.Wrapf(sdkerrors.ErrPanic, "%v", r))
+		}
+	}()
+
+	// when a client did not provide a query height, manually inject the latest
+	if req.Height == 0 {
+		req.Height = s.LastBlockHeight()
+	}
+
+	path := SplitABCIQueryPath(req.Path)
+	if len(path) == 0 {
+		return queryResult(errorsmod.Wrap(sdkerrors.ErrUnknownRequest, "no query path provided")), nil
+	}
+
+	switch path[0] {
+	case "store":
+		// "/store" prefix for store queries
+		req.Path = "/" + strings.Join(path[1:], "/")
+		cms := s.sm.GetCommitMultiStore()
+		queryable, ok := cms.(storetypes.Queryable)
+		if !ok {
+			panic(ok)
```

Comment:
might be better to return error here like [cosmos](https://github.com/cosmos/cosmos-sdk/blob/960d44842b9e313cbe762068a67a894ac82060ab/baseapp/abci.go#L1102) rather than panic

Reply:
ah this was just from our testing, forgot to copy the actual error handling in. will fix 👍